File: apps/server_py/app/domain/session/session_stream_lock.py
from __future__ import annotations

import asyncio
import os
import time
from typing import Awaitable, Callable, Optional

from app.config import get_settings
from app.domain.session.errors import SessionStreamLockBusyError
from app.domain.session.get_repository import resolve_session_store_mode
from app.infra.redis_client import get_redis_client

LOCK_KEY_PREFIX = "kronos:session:lock:"
DEFAULT_LOCK_TTL_SEC = 120

ReleaseLock = Callable[[], Awaitable[None]]


def _is_stream_lock_enabled() -> bool:
    if resolve_session_store_mode() != "redis":
        return False
    raw = (get_settings().session_stream_lock or "true").strip().lower()
    return raw not in ("false", "0", "no")


def _resolve_lock_ttl_sec() -> int:
    raw = get_settings().session_stream_lock_ttl_sec
    return raw if raw > 0 else DEFAULT_LOCK_TTL_SEC


async def acquire_session_stream_lock(session_id: str) -> Optional[ReleaseLock]:
    if not _is_stream_lock_enabled():
        return None

    client = get_redis_client()
    key = f"{LOCK_KEY_PREFIX}{session_id}"
    token = f"{os.getpid()}-{time.time()}"
    acquired = await client.set(key, token, ex=_resolve_lock_ttl_sec(), nx=True)

    if not acquired:
        raise SessionStreamLockBusyError(session_id)

    async def release() -> None:
        current = await client.get(key)
        if current == token:
            await client.delete(key)

    return release

File: apps/server/src/models/session/sessionStreamLock.ts
import { duplicateRedisClient } from '../infra/redisClient.js';
import { resolveSessionStoreMode } from './getSessionRepository.js';

const LOCK_KEY_PREFIX = 'kronos:session:lock:';
const DEFAULT_LOCK_TTL_SEC = 120;

export class SessionStreamLockBusyError extends Error {
  readonly code = 'SESSION_STREAM_BUSY';

  constructor(readonly sessionId: string) {
    super(`Session ${sessionId} is locked by another stream`);
    this.name = 'SessionStreamLockBusyError';
  }
}

const resolveLockTtlSec = (): number => {
  const raw = Number(process.env.SESSION_STREAM_LOCK_TTL_SEC ?? DEFAULT_LOCK_TTL_SEC);
  return Number.isFinite(raw) && raw > 0 ? Math.floor(raw) : DEFAULT_LOCK_TTL_SEC;
};

const isStreamLockEnabled = (): boolean => {
  if (resolveSessionStoreMode() !== 'redis') {
    return false;
  }

  const raw = (process.env.SESSION_STREAM_LOCK ?? 'true').trim().toLowerCase();
  return raw !== 'false' && raw !== '0' && raw !== 'no';
};

/** Redis 模式下串行同 session 的 SSE，避免双开流写冲突。 */
export const acquireSessionStreamLock = async (
  sessionId: string,
): Promise<(() => Promise<void>) | null> => {
  if (!isStreamLockEnabled()) {
    return null;
  }

  const redis = duplicateRedisClient();
  const key = `${LOCK_KEY_PREFIX}${sessionId}`;
  const token = `${process.pid}-${Date.now()}`;
  const acquired = await redis.set(key, token, 'EX', resolveLockTtlSec(), 'NX');

  if (acquired !== 'OK') {
    redis.disconnect();
    throw new SessionStreamLockBusyError(sessionId);
  }

  return async () => {
    try {
      const current = await redis.get(key);
      if (current === token) {
        await redis.del(key);
      }
    } finally {
      redis.disconnect();
    }
  };
};
